tests/test_gpu_examples.py

```diff
@@ -933,3 +936,49 @@ def test_causal_lm_training_mutli_gpu(self):
 
             # assert loss is not None
             self.assertIsNotNone(trainer.state.log_history[-1]["train_loss"])
+
+
+@require_torch_gpu
+class OffloadSaveTests(unittest.TestCase):
+    def setUp(self):
+        self.causal_lm_model_id = "gpt2"
+
+    def tearDown(self):
+        r"""
+        Efficient mechanism to free GPU memory after each test. Based on
+        https://github.com/huggingface/transformers/issues/21094
+        """
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @pytest.mark.single_gpu_tests
+    @require_torch_gpu
+    def test_offload_merge(self):
+        r"""
+        Test merging and unloading of a model with offloaded modules.
+        """
+        tokenizer = AutoTokenizer.from_pretrained(self.causal_lm_model_id)
+        config = AutoConfig.from_pretrained(self.causal_lm_model_id)
+        with init_empty_weights():
+            model = AutoModelForCausalLM.from_config(config)
+
+        memory_limits = {0: "0.5GIB", "cpu": "5GIB"}
+        # offloads around half of all transformer modules
+        device_map = infer_auto_device_map(model, max_memory=memory_limits)
+        assert 0 in device_map.values()
+        assert "cpu" in device_map.values()
+
+        model = AutoModelForCausalLM.from_pretrained(self.causal_lm_model_id, device_map=device_map)
+        assert len({p.device for p in model.parameters()}) == 2
+
+        # LoRA adapters on ['c_attn', 'c_proj', 'c_fc']
+        lora_checkpoint = "blbadger/gpt2-lora"
```

Comment:
Thinking more about this, do we even need a lora checkpoint on the Hub? I believe it would be easier to create a LoRA model on the fly:

```python
        torch.manual_seed(0)
        config = LoraConfig(
            task_type="CAUSAL_LM",
            init_lora_weights=False,
        )
        model = AutoModelForCausalLM.from_pretrained(self.causal_lm_model_id)
        model = get_peft_model(model, config)

        with tempfile.TemporaryDirectory() as tmp_dir:
            model.save_pretrained(tmp_dir)
            # load the model with device_map
            model = AutoModelForCausalLM.from_pretrained(self.causal_lm_model_id, device_map=device_map).eval()
            self.assertTrue(len({p.device for p in model.parameters()}) == 2)
            model = PeftModel.from_pretrained(model, tmp_dir, max_memory=memory_limits)
```

WDYT?

Reply:
This sounds good to me! As long as we load via `.from_pretrained` the test should work well (once we get the merging accuracy fixed that is)